fission-web-server/library/Fission/Web/Server/Types.hs

```diff
@@ -397,15 +325,53 @@ instance IPFS.MonadLocalIPFS Server where
 
 instance IPFS.MonadRemoteIPFS Server where
   runRemote query = do
-    peerID       <- head <$> asks ipfsRemotePeers
-    IPFS.URL url <- asks ipfsURL
-    manager      <- asks httpManager
+    clusterURLs       <- asks ipfsURLs
+    IPFS.Timeout secs <- asks ipfsTimeout
+    manager           <- asks httpManager
+
+    logDebug @Text "Running IPFS request across cluster"
+    requests <- forM clusterURLs \(IPFS.URL url) ->
+      Process.asyncFor (Seconds (Unity secs)) do
+        runClientM query $ mkClientEnv manager url
 
-    Peer.connectRetry peerID 25 >>= \case
-      Left  err -> logDebug $ "Unable to connect: " <> textDisplay err
-      Right _   -> return ()
+    liftIO $ untilDone requests
 
-    liftIO . runClientM query $ mkClientEnv manager url
+    where
+      untilDone ::
+           NonEmpty (Async (Either Process.TimedOut (Either ClientError a)))
+        -> IO (Either ClientError a)
+      untilDone asyncRefs = do
+        results <- sequence (getStatus <$> asyncRefs)
+        case Process.progress results of
+          Process.Failed err -> return $ Left err
+          Process.InProgress -> untilDone asyncRefs
+          Process.Success a  -> return $ Right a
+      -- NOTE letting GHC's scheduler handle the frequency of retries.
+      -- This should be pretty lightweight, since we're just waiting for processes to return.
+      -- That said, if we go to 99% CPU or something, this is probably a good place to look 😉
+
+      getStatus ::
+           Async (Either Process.TimedOut (Either ClientError a))
+        -> IO    (Process.Status ClientError a)
+      getStatus asyncThread =
+        poll asyncThread >>= \case
+          Nothing -> -- 🏋️‍ Still working on it
+            return Process.InProgress
+
+          Just result ->
+            case result of
+              Left procError -> -- 🧨 Inner action threw an exception
+                return . Process.Failed $ ConnectionError procError
+
+              Right maybeTimedOut ->
+                case maybeTimedOut of
+                  Left Process.TimedOut -> -- ⏲️  Timed out
+                    return . Process.Failed . ConnectionError $ toException Process.TimedOut
+
+                  Right requestResult ->
+                    case requestResult of -- 🌐 HTTP request result
+                      Left  clientErr -> return $ Process.Failed  clientErr
+                      Right a         -> return $ Process.Success a
```

Comment:
Lots of layers happening in that `Async` 😛 This is a consequence of us wanting a first-past-the-post strategy. A `race` would have been much simpler, but we _want_ the rest of the requests to continue while we report success to the user. Which means we need to add a timeout so they don't run literally forever in the degenerate case.

1. The `Async` itself may not be done (`poll :: Async a -> IO (Maybe a)`)
2. The inner process may throw an IO-based exception (`Either SomeException b`)
3. We have a timeout watcher, so that's an `Either Process.Timeout c`
4. Finally, the result of the HTTP request — `Either ClientError d`